MoreLinq.NoConflictGenerator/Program.cs

```diff
@@ -0,0 +1,455 @@
+
+            var qq =
+                from fp in Directory.EnumerateFiles(dir, "*.cs")
+                where !excludePredicate(fp) && includePredicate(fp)
+                orderby fp
+                from m in
+                    from cd in
+                        CSharpSyntaxTree
+                            .ParseText(File.ReadAllText(fp), CSharpParseOptions.Default.WithPreprocessorSymbols("MORELINQ"))
+                            .GetRoot()
+                            .SyntaxTree
+                            .GetCompilationUnitRoot()
+                            .DescendantNodes().OfType<ClassDeclarationSyntax>()
+                    where (string) cd.Identifier.Value == "MoreEnumerable"
+                    from md in cd.DescendantNodes().OfType<MethodDeclarationSyntax>()
+                    let mn = (string) md.Identifier.Value
+                    where md.ParameterList.Parameters.Count > 0
+                       && md.ParameterList.Parameters.First().Modifiers.Any(m => (string)m.Value == "this")
+                       && md.Modifiers.Any(m => (string)m.Value == "public")
+                       && md.AttributeLists.SelectMany(al => al.Attributes).All(a => a.Name.ToString() != "Obsolete")
+                    let typeParameterAbbreviationByName =
+                        md.TypeParameterList
+                         ?.Parameters
+                          .Select((e, i) => (Original: e.Identifier.ValueText, Alias: abbreviatedTypeNodes[i]))
+                          .ToDictionary(e => e.Original, e => e.Alias)
+                    select new
+                    {
+                        Syntax = md,
+                        Name = md.Identifier.ToString(),
+                        TypeParameterCount = md.TypeParameterList?.Parameters.Count ?? 0,
+                        TypeParameterAbbreviationByName = typeParameterAbbreviationByName,
+                        ParameterCount = md.ParameterList.Parameters.Count,
+                        SortableParameterTypes =
+                            from p in md.ParameterList.Parameters
+                            select CreateTypeKey(p.Type,
+                                                 n => typeParameterAbbreviationByName != null
+                                                   && typeParameterAbbreviationByName.TryGetValue(n, out var a) ? a : null),
+                    }
+                select m;
+
+        var q =
+            from e in qq.Select((m, i) => (SourceOrder: i + 1, Method: m))
+            orderby
+                e.Method.Name,
+                e.Method.TypeParameterCount,
+                e.Method.ParameterCount,
+                new TupleTypeKey(ImmutableList.CreateRange(e.Method.SortableParameterTypes))
+            select new
+            {
+                e.Method,
+                e.SourceOrder,
+            };
+
+            q = q.ToArray();
+
+            if (debug)
+            {
+                foreach (var m in
+                    from e in q
```

Comment:
I don't think I'm convinced by this query-inside-foreach. I think just unfolding it into the loop would be better since there is no filtering going on. Alternatively, putting the query into a temporary variable would work too.

Reply:
Programming in expressions (the query inside) provides more clarity and opportunities for re-use. More importantly, it helps separate data transformation and queries from actual side-effects where you act on the results, and which is delayed until the body of the for-loop. In performance-sensitive code, I would do as you suggest here. I realise this may seem a bit extreme, but believe me, it saves you cognitive tax on coming up with artificially unique variable names (that often end up being long-winded) for each intermediate representation on your journey to the final result. The query syntax is the only part of C# (besides artificial blocks), today, that allows reusing or re-typing names through _range variables_ (e.g. `m` and `e` here).

> Alternatively, putting the query into a temporary variable would work too

I can do this. Now I have to come up with a name. 😖